### DISCUSSION TITLE
How to use GTM in nuxt3

### DISCUSSION THREAD

**Discussion by KahrLu** [nuxt/nuxt#17717] [Questions] [answered]
Hi @pi0 , the two plugins of gtm in nuxt2 doesn‘t work in nuxt3! sad....
and also I tried to use [vue-gtm](https://github.com/gtm-support/vue-gtm) in nuxt, failed either....
see https://github.com/gtm-support/vue-gtm/issues/178
any other way to use GTM in nuxt3 successfully?
could u help me on this , thanks!

**Comment by oritwoen** [comment]
Check this: https://v3.nuxtjs.org/guide/directory-structure/plugins/#vue-plugins

**Comment by benoitdemaegdt** [comment, marked as answer]
Hey @KahrLu  👋 

There are several options.

1/ Try to install it with partytown module
=> https://github.com/nuxt-community/partytown-module#google-tag-manager
However, you should be aware that the partytown module does not work with latest nuxt3 version
=> https://github.com/nuxt-community/partytown-module/issues/58
The PR that should fix it is open but not yet merged
=> https://github.com/nuxt/framework/pull/4304

_EDIT : partytown module has been fixed_ 🎉 

2/ Install it manually
This is how I personally installed GTM on my nuxt3 app. I hope that I'll soon be able to change this, because it is quite complicated.

2.1/ Install vue-gtm
`npm i @gtm-support/vue-gtm` 

2.2/ create a gtm plugin
`/plugins/vue-gtm.client.js`

Then copy/paste this code inside this file
```js
import { createGtm } from '@gtm-support/vue-gtm'

export default defineNuxtPlugin(nuxtApp => {
  nuxtApp.vueApp.use(
    createGtm({
      id: <YOUR_GTM_ID>,
      defer: false, // Script can be set to `defer` to speed up page load at the cost of less accurate results (in case visitor leaves before script is loaded, which is unlikely but possible). Defaults to false, so the script is loaded `async` by default
      compatibility: false, // Will add `async` and `defer` to the script tag to not block requests for old browsers that do not support `async`
      // nonce: '2726c7f26c', // Will add `nonce` to the script tag
      enabled: true, // defaults to true. Plugin can be disabled by setting this to false for Ex: enabled: !!GDPR_Cookie (optional)
      debug: true, // Whether or not display console logs debugs (optional)
      loadScript: true, // Whether or not to load the GTM Script (Helpful if you are including GTM manually, but need the dataLayer functionality in your components) (optional)
      trackOnNextTick: false, // Whether or not call trackView in Vue.nextTick
    })
  )
})
```
 
2.3/ After doing this, I noticed that the pageView Trigger from GTM wasn't really working. Sometime it gets the previous URL, sometime it doesn't get anything at all. I think GTM PageViews are triggered _before_ the actual route change ...
So I ended pushing manually the page view to GTM. To do this I created a new middleware
`/middleware/pageView.global.js`
And added this code inside : 
```js
export default defineNuxtRouteMiddleware((to) => {
  if (process.server) return
  sendGtmPageView({ pagePath: to.path , pageTitle: to.name })
})

function sendGtmPageView({ pagePath, pageTitle }) {
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({ event: 'Pageview', pagePath, pageTitle });
}
```

With all of this, your GTM account should be working with Nuxt3 😄
